### ISSUE TITLE
Custom Serialization Should Use `fromJson` Parameter Type Instead of Fixed `Map<String, dynamic>`

### ISSUE THREAD

**Issue by AndryHTC** [serverpod/serverpod#2715]
#### Description
Currently, the `fromJson` method signature in the generated code uses `Map<String, dynamic>` as a fixed type for serialization:

```dart
factory CustomSerializableModel.fromJson(Map<String, dynamic> jsonSerialization) {
    return CustomSerializableModel(
      uuid: jsonSerialization['uuid'] as String,
      enumField: _i2.CustomSerializableEnum.fromJson(
          (jsonSerialization['enumField'] as Map<String, dynamic>)),
    );
  }
```

This implementation expects `enumField` to be deserialized from a `Map<String, dynamic>`, which may be adequate in most of simple cases but is overly restrictive in more advanced scenarios. When using **Enhanced Enums** or other custom serialization logic, there are situations where we only need to work with a **string** as input or output for serialization, rather than a complex object.

#### Problem
For instance, in my case, I'm serializing an **Enhanced Enum** that is shared between a server and a client. The server returns a simple **string** (e.g., `"ENUM_VALUE"`) rather than a complex object. I’ve tried workarounds like wrapping the value in a map:

```dart
{"name": name}
```

However, this approach leads to unnecessarily bloated data and makes the `fromJson` method incompatible with simpler custom serializations, such as the one used by Dart ORM, which logically returns a **string** for an enum instead of a `Map`.

The current implementation expects a map-like structure for `enumField`, which causes deserialization failures when the input is not a `Map<String, dynamic>`, even though the actual data could be represented more efficiently.

#### Proposal
The method should be updated to accept a more flexible parameter type for `fromJson`, such as `T fromJson(dynamic json)`. This would allow developers to pass and return data types suited to their specific use case, including strings or other simple structures. This small change would improve flexibility and avoid the bloating caused by unnecessarily complex data structures in serialization.

#### Use Case
In my current project, I am serializing Enhanced Enums and would like to return a **string** rather than a full `Map<String, dynamic>`. This would prevent inflated data size and maintain compatibility with other custom serializations, such as Dart ORM’s, which logically expects a string representation for enums.

By loosening the restrictions on the input type of `fromJson`, this flexibility would allow for much broader use cases and provide a more efficient serialization process.

#### PLEASE
I really hope to have this in 2.1 as I'm migrating a hundred users production app for the next release.
It's a no-brainer, as is not Serverpod responsibility but the dev's one to keep the `fromJson` and `toJson` data type compatible.

**Comment by SandPod** [comment]
This makes total sense.

Although I do believe a `fromJson` and `toJson` interface should actually parse a well-formed JSON represented by `Map<String, dynamic>` in Dart, Serverpod should not be opinionated on the matter.

@AndryHTC, please feel free to submit a PR to fix this, and I will make sure to review it. Simply removing `as Map<String, dynamic>` in the generated code should be sufficient to fix the issue.

(I also noticed that our enum implementation is not correct. The `fromJson` and `toJson` methods return a simple value instead of a well formed JSON 🙃, sadly this is a breaking change that we will have to revisit for Serverpod 3.0).

**Comment by AndryHTC** [comment]
Thanks for the quick reply, @SandPod!

I completely agree that both `fromJson` and `toJson` should handle valid JSON representations, and it makes sense to adhere to the `Map<String, dynamic>` structure for consistency. However, as a temporary fix, removing the strict type in the generated code will unblock custom implementations like mine that don’t strictly require a `Map`. 

As for a longer-term solution, I think introducing more generic methods such as `fromSerialized`/`toSerialized` or `parse`/`serialize` could provide better flexibility for different serialization formats. This would avoid confusion and still allow for the necessary separation between JSON parsing and more custom or lightweight serialization.

I’ll proceed with the temporary fix for now and submit a PR soon. Thanks again for the feedback!

**Comment by AndryHTC** [comment]
@SandPod I appreaciate very much your promptness, but I didn't succeded even on testing the changes. If is a change so simple, please, consider to do that on my behalf.

More on my abortion [here](https://github.com/serverpod/serverpod/discussions/2067)

**Comment by SandPod** [comment]
We will try to prioritize this issue.

It might not make it to the initial release of Serverpod 2.1.0, but we can probably include it in a patch fix following shortly thereafter.

**Comment by AndryHTC** [comment]
> @AndryHTC a fix for this has now been merged that will be included in Serverpod 2.1.1 that will be published in a couple of hours.

That's great. Thank you for keeping me posted
